File: Ch1/Java/RotateMatrix.java
import java.util.Arrays;
public class RotateMatrix {

    public static void main(String[] args) {
        int[][] matrix = new int[8][8];
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                matrix[i][j] = (int) (Math.random()*10) + 1;
            }
        }
        print2DArray(matrix);
        rotateMatrix(matrix);
        print2DArray(matrix);
    }
    
    public static void print2DArray(int[][] arr) {
        System.out.printf("%s\n\n", Arrays.deepToString(arr).replace("], ", "]\n"));
    }

    public static void rotateMatrix(int[][] matrix) {
        //rotate's matrix clockwise 90 degrees in-place
        int[][] newIndexes = new int[matrix.length * matrix.length][1];
        int index = 0;
        for (int row = 0; row < matrix.length; row++) {
            for (int col = 0; col < matrix[row].length; col++) {
                // System.out.printf("(%d, %d): %d\n", row, col, matrix[row][col]);
                int newRow = col;
                int newCol = matrix.length - row - 1;
                newIndexes[index] = new int[]{newRow, newCol, matrix[row][col]};
                index++;
            }
        }
        for (int[] mapping : newIndexes) {
            int newRow = mapping[0], newCol = mapping[1], val = mapping[2];
            matrix[newRow][newCol] = val;
        }
    }
}

File: Ch1/C++/RotateMatrix.cpp
#include <stdlib.h>

#include <iostream>
#include <vector>

void print2DVector(const std::vector<std::vector<int>>& matrix) {
    for (std::vector<int> row : matrix) {
        std::cout << "[";
        for (int i = 0; i < row.size(); i++) {
            if (i != row.size() - 1) {
                std::cout << row.at(i) << ", ";
            } else {
                std::cout << row.at(i);
            }
        }
        std::cout << "]" << std::endl;
    }
    std::cout << std::endl;
}

void rotateMatrix(std::vector<std::vector<int>>& matrix) {
    std::vector<std::vector<int>> newIndexes(matrix.size() * matrix.size(), std::vector<int>(1, 0));
    int index = 0;
    for (int row = 0; row < matrix.size(); row++) {
        for (int col = 0; col < matrix[row].size(); col++) {
            int newRow = col;
            int newCol = matrix.size() - row - 1;
            newIndexes[index] = std::vector<int>{newRow, newCol, matrix[row][col]};
            index++;
        }
    }
    for (std::vector<int> mapping : newIndexes) {
        int newRow = mapping[0], newCol = mapping[1], val = mapping[2];
        matrix[newRow][newCol] = val;
    }
}

int main() {
    std::vector<std::vector<int>> matrix(8, std::vector<int>(8, 0));
    for (int i = 0; i < matrix.size(); i++) {
        for (int j = 0; j < matrix[i].size(); j++) {
            matrix[i][j] = rand() % 10 + 1;
        }
    }
    print2DVector(matrix);
    rotateMatrix(matrix);
    print2DVector(matrix);
    return 0;
};